eventbus/kafka/sensor/kafka_sensor.go

```diff
@@ -0,0 +1,367 @@
+	return !s.connected || s.client.Closed()
+}
+
+func (s *KafkaSensor) Event(msg *sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()) {
+	var event *cloudevents.Event
+	if err := json.Unmarshal(msg.Value, &event); err != nil {
+		s.Logger.Errorw("Failed to deserialize cloudevent, skipping", zap.Error(err))
+		return nil, msg.Offset + 1, nil
+	}
+
+	messages := []*sarama.ProducerMessage{}
+	for _, trigger := range s.triggers.List(event) {
+		event, err := trigger.Transform(trigger.depName, event)
+		if err != nil {
+			s.Logger.Errorw("Failed to transform cloudevent, skipping", zap.Error(err))
+			continue
+		}
+
+		if !trigger.Filter(trigger.depName, event) {
+			s.Logger.Debug("Filter condition satisfied, skipping")
+			continue
+		}
+
+		// if the trigger only requires one message to be invoked we
+		// can skip ahead to the action topic, otherwise produce to
+		// the trigger topic
+
+		var data any
+		var topic string
+		if trigger.OneAndDone() {
+			data = []*cloudevents.Event{event}
+			topic = s.topics.action
+		} else {
+			data = event
+			topic = s.topics.trigger
+		}
+
+		value, err := json.Marshal(data)
+		if err != nil {
+			s.Logger.Errorw("Failed to serialize cloudevent, skipping", zap.Error(err))
+			continue
+		}
+
+		messages = append(messages, &sarama.ProducerMessage{
+			Topic: topic,
+			Key:   sarama.StringEncoder(trigger.Name()),
+			Value: sarama.ByteEncoder(value),
+		})
+	}
+
+	return messages, msg.Offset + 1, nil
+}
+
+func (s *KafkaSensor) Trigger(msg *sarama.ConsumerMessage) ([]*sarama.ProducerMessage, int64, func()) {
+	var event *cloudevents.Event
+	if err := json.Unmarshal(msg.Value, &event); err != nil {
+		// do not return here as we still need to call trigger.Offset
+		// below to determine current offset
+		s.Logger.Errorw("Failed to deserialize cloudevent, skipping", zap.Error(err))
+	}
+
+	messages := []*sarama.ProducerMessage{}
+	offset := msg.Offset + 1
+
+	// update trigger with new event and add any resulting action to
+	// transaction messages
+	if trigger, ok := s.triggers[string(msg.Key)]; ok && event != nil {
+		func() {
+			events, err := trigger.Update(event, msg.Partition, msg.Offset, msg.Timestamp)
+			if err != nil {
+				s.Logger.Errorw("Failed to update trigger, skipping", zap.Error(err))
+				return
+			}
+
+			// no events, trigger not yet satisfied
+			if events == nil {
+				return
+			}
+
+			value, err := json.Marshal(events)
+			if err != nil {
+				s.Logger.Errorw("Failed to serialize cloudevent, skipping", zap.Error(err))
+				return
+			}
+
+			messages = append(messages, &sarama.ProducerMessage{
+				Topic: s.topics.action,
+				Key:   sarama.StringEncoder(trigger.Name()),
+				Value: sarama.ByteEncoder(value),
+			})
+		}()
+	}
+
+	// need to determine smallest possible offset against all
+	// triggers as other triggers may have messages that land on the
+	// same partition
+	for _, trigger := range s.triggers {
+		offset = trigger.Offset(msg.Partition, offset)
+	}
+
```

Comment:
can this handler be called by two threads at the same time or no? if so, any issues related to both threads executing lines 347-349 at the same time?

Reply:
No, luckily this is not possible by construction. We use the kafka message key to look up the trigger [here](https://github.com/argoproj/argo-events/pull/2502/files#diff-b778225ecfb035ee9a44077f3f987fd20bc3381f1786485f86639419b0c9196fR317), a key always maps to only one partition and all messages pertaining to the same partition are processed sequentially (and by one pod). That means we don't have to worry about concurrency in our stateful code (mostly the KafkaTriggerConnection). This is also how we enable scaling - all events that pertain to a specific trigger will land on the same partition and are guaranteed to be processed by the same pod.

Reply:
Right, I guess my confusion was that we were looking at other triggers outside of our own on line 347

Reply:
I think I see what you're saying. The Offset() function only goes through events on this same partition, and all messages pertaining to the same partition are processed sequentially.